backend/core/src/shard/connector.rs

```diff
@@ -0,0 +1,156 @@
+use std::mem;
+use std::time::{Duration, Instant};
+
+use crate::aggregator::{AddNode, Aggregator};
+use crate::chain::{Chain, RemoveNode, UpdateNode};
+use crate::shard::ShardMessage;
+use crate::types::NodeId;
+use crate::util::{DenseMap, Hash};
+use actix::prelude::*;
+use actix_http::ws::Item;
+use actix_web_actors::ws::{self, CloseReason};
+use bincode::Options;
+use bytes::{Bytes, BytesMut};
+
+/// How often heartbeat pings are sent
+const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(20);
+/// How long before lack of client response causes a timeout
+const CLIENT_TIMEOUT: Duration = Duration::from_secs(60);
+/// Continuation buffer limit, 10mb
+const CONT_BUF_LIMIT: usize = 10 * 1024 * 1024;
+
+pub struct ShardConnector {
+    /// Client must send ping at least once every 60 seconds (CLIENT_TIMEOUT),
+    hb: Instant,
+    /// Aggregator actor address
+    aggregator: Addr<Aggregator>,
+    /// Genesis hash of the chain this connection will be submitting data for
+    genesis_hash: Hash,
+    /// Chain address to which this multiplex connector is delegating messages
+    chain: Option<Addr<Chain>>,
+    /// Mapping `ShardConnId` to `NodeId`
+    nodes: DenseMap<NodeId>,
+    /// Buffer for constructing continuation messages
+    contbuf: BytesMut,
+}
+
+impl Actor for ShardConnector {
+    type Context = ws::WebsocketContext<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        self.heartbeat(ctx);
+    }
+
+    fn stopped(&mut self, _: &mut Self::Context) {
+        if let Some(ref chain) = self.chain {
+            for (_, nid) in self.nodes.iter() {
+                chain.do_send(RemoveNode(*nid))
+            }
+        }
+    }
+}
+
+impl ShardConnector {
+    pub fn new(aggregator: Addr<Aggregator>, genesis_hash: Hash) -> Self {
+        Self {
+            hb: Instant::now(),
+            aggregator,
+            genesis_hash,
+            chain: None,
+            nodes: DenseMap::new(),
+            contbuf: BytesMut::new(),
```

Comment:
Per the docs for `ByteMut::new()`

> Creates a new BytesMut with default capacity.
> 
> Resulting object has length 0 and unspecified capacity. This function does not allocate.

I am assuming that `allocate` just means that `BytesMut` does not use any dynamic memory handling? Does that mean everything is just on the active stack?

Reply:
Behind the scenes `Bytes` and `BytesMut` are backed by a `Vec`, which uses heap memory managed by an allocator. So it **does** use dynamic memory handling. What the docs are pointing out is that `new()` does not pre-allocate a buffer for future use (if you want that, use `::with_capacity()`). Check out [the source](https://docs.rs/bytes/1.0.1/src/bytes/bytes_mut.rs.html#165-167) to see for yourself! :)

Reply:
TIL: You can look at the source via the docs. Incredibly cool.

 Yea that definitely answers my question! Thank you :)

Reply:
> TIL: You can look at the source via the docs. Incredibly cool.

Yeah once you get used to the perks of rustdocs it's hard to not miss it in other languages.